crates/tags/src/tag.rs

```diff
@@ -1,60 +1,241 @@
 // SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
 // SPDX-License-Identifier: Apache-2.0
 
-use std::hash::Hash;
 use std::ops::{Deref, DerefMut};
 use std::str::FromStr;
 
-use drawbridge_http::http::{self, Error, Request, StatusCode};
+use drawbridge_http::http::Request;
 use drawbridge_http::{async_trait, FromRequest};
+use drawbridge_jose::jws::Jws;
+use drawbridge_type::Entry;
 
-use semver::{self, Version};
 use serde::{Deserialize, Serialize};
 
-#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
-pub struct Tag(Version);
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct TagName(String);
 
-impl Deref for Tag {
-    type Target = Version;
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct TagValue {
+    pub body: Vec<u8>,
+    pub kind: Kind,
+    pub name: TagName,
+}
+
+impl TagValue {
+    pub const TYPE_ENTRY: &'static str = "application/vnd.drawbridge.tag.v1+json";
+    pub const TYPE_JOSE: &'static str = "application/jose+json";
+}
+
+// value of the HashMap, name is the Key
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum Kind {
+    Signed(Jws),
+    Unsigned(Entry),
+}
+
+impl Deref for TagName {
+    type Target = String;
 
     fn deref(&self) -> &Self::Target {
         &self.0
     }
 }
 
-impl DerefMut for Tag {
+impl DerefMut for TagName {
     fn deref_mut(&mut self) -> &mut Self::Target {
         &mut self.0
     }
 }
 
-impl FromStr for Tag {
-    type Err = semver::Error;
+impl FromStr for TagName {
+    type Err = ();
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        s.parse().map(Self)
+        Ok(TagName(s.to_string()))
     }
 }
 
 #[async_trait]
-impl FromRequest for Tag {
-    async fn from_request(req: &mut Request) -> http::Result<Self> {
-        req.url().path()[1..]
-            .parse()
-            .map_err(|_| Error::from_str(StatusCode::BadRequest, ""))
+impl FromRequest for TagName {
+    async fn from_request(req: &mut Request) -> http_types::Result<Self> {
+        // todo: properly get the tag name from the url
+        // if not a fully-qualified URL, the tests will fail
+        let tag_name = match req.url().as_str().split("/").last() {
+            None => {
+                return Err(http_types::Error::from_str(
+                    http_types::StatusCode::BadRequest,
+                    "invalid URL",
+                ))
+            }
+            Some(x) => x,
+        };
```

Comment:
```suggestion
        let tag_name = match req.url().path().strip_prefix('/').expect("invalid URI").split('/').next().ok_or(Error::from_str(
                    StatusCode::BadRequest,
                    "Tag must be specified",
                ));
```

Reply:
Can't use `match` here, and there's nothing being matched. And I don't know how to handle `Result<&str,Error>` to make it `Result<Name,Error>`. This line is not readable.

Reply:
sorry, there should not be a match there and you should run `cargo fmt` on it